### ISSUE TITLE
Enable more deep learning tests

### ISSUE THREAD

**Issue by dkurt** [opencv/opencv#11866]
##### System information (version)

- OpenCV => :grey_question:
- Operating System / Platform => :grey_question:
- Compiler => :grey_question:

##### Detailed description

The following tests are skipped using specific backend+target configurations. Classify them by categories (breaks with an assertion :negative_squared_cross_mark:, accuracy loss :beetle:), find a way to fix (as a bug solution or a feature request :star: ). Some of tests could be fixed due Inference Engine graphs building fix (it's assigned to @dkurt).

* DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16

- [ ] Test_TensorFlow_layers.lstm :beetle: 
- [ ] Test_TensorFlow_layers.matmul (on [buildbot](http://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux)) :beetle: 
- [ ] Test_Caffe_layers.InnerProduct :beetle:
- [ ] Test_Caffe_layers.layer_prelu_fc :beetle: (unstable)
- [ ] Test_Caffe_layers.FasterRCNN_Proposal :beetle: 
- [x] Test_Caffe_layers.PriorBox_squares :beetle: 
- [x] Layer_Test_Halide/Convolution.Accuracy,([6, 4, 1], 5x6, 3x1, 1x1, 0x1, 1x1, false) :beetle: (unstable)
- [x] Layer_Test_Halide/Convolution.Accuracy,([6, 4, 1], 5x6, 3x1, 1x1, 0x1, 2x2, true) :beetle: (unstable)
- [x] Layer_Test_Halide/FullyConnected.Accuracy :negative_squared_cross_mark:
 
* DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL

- [x] Test_Caffe_layers.PriorBox_squares :beetle: 

* DNN_BACKEND_INFERENCE_ENGINE, DNN_TARGET_CPU

- [ ] Test_Darknet_layers.shortcut :negative_squared_cross_mark:
- [x] Test_TensorFlow_layers.mvn_batch_norm :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.reshape :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.lstm :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.split :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.resize_nearest_neighbor :beetle: 
- [x] Test_Caffe_layers.DeConvolution :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.InnerProduct :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.BatchNorm :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Fused_Concat :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Eltwise :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Reshape_Split_Slice :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.FasterRCNN_Proposal :star:
- [ ] Test_Caffe_layers.PriorBox_squares :negative_squared_cross_mark:

* DNN_BACKEND_INFERENCE_ENGINE, DNN_TARGET_OPENCL

- [ ] Test_TensorFlow_layers.reshape :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.flatten :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.lstm :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.split :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.resize_nearest_neighbor :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.slice :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.InnerProduct :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Fused_Concat :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Eltwise :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Reshape_Split_Slice :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.FasterRCNN_Proposal :star:
- [ ] Test_Caffe_layers.PriorBox_squares :negative_squared_cross_mark:
- [x] Test_TensorFlow_layers.mvn_batch_norm :negative_squared_cross_mark:
- [x] Test_Caffe_layers.DeConvolution :beetle:
- [x] Test_Caffe_layers.BatchNorm :negative_squared_cross_mark: (enabled starts from R3)

* DNN_BACKEND_INFERENCE_ENGINE, DNN_TARGET_OPENCL_FP16

- [ ] Test_TensorFlow_layers.reshape :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.flatten :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.lstm :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.split :beetle: 
- [ ] Test_TensorFlow_layers.resize_nearest_neighbor :negative_squared_cross_mark:
- [ ] Test_TensorFlow_layers.slice :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.InnerProduct :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Eltwise :beetle:
- [ ] Test_Caffe_layers.Reshape_Split_Slice :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.FasterRCNN_Proposal :star:
- [ ] Test_Caffe_layers.PriorBox_squares :negative_squared_cross_mark:
- [x] Test_Caffe_layers.BatchNorm :negative_squared_cross_mark: (enabled starts from R3)
- [x] Test_TensorFlow_layers.mvn_batch_norm :negative_squared_cross_mark:
- [x] Test_Caffe_layers.DeConvolution :beetle:

* DNN_BACKEND_INFERENCE_ENGINE, DNN_TARGET_MYRIAD

- [ ] Test_TensorFlow_layers.l2_normalize_3d :beetle:
- [ ] Test_Caffe_layers.LRN_spatial :beetle:
- [ ] Test_Caffe_layers.InnerProduct :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Eltwise :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Reshape_Split_Slice :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.FasterRCNN_Proposal :star:
- [ ] Test_Caffe_layers.PriorBox_squares :negative_squared_cross_mark:
- [ ] Test_Caffe_layers.Interp :negative_squared_cross_mark:
- [x] Test_TensorFlow_layers.fp16_pad_and_concat :beetle: (enabled starts from R3)
- [x] Test_TensorFlow_layers.ave_pool_same :beetle: (enabled starts from R3)
- [x] Test_TensorFlow_layers.mvn_batch_norm :negative_squared_cross_mark:
- [x] Test_Caffe_layers.DeConvolution :beetle:
- [x] Test_TensorFlow_layers.pad_and_concat :beetle: (enabled starts from R3)
- [x] Test_Caffe_layers.BatchNorm :negative_squared_cross_mark: (enabled starts from R3)
- [x] Test_Caffe_layers.Average_pooling_kernel_area :beetle: (enabled starts from R3)

##### Steps to reproduce

Build OpenCV with `ts` module and tests enabled:

```bash
cmake -DCMAKE_BUILD_TYPE=RELEASE -DBUILD_TESTS=ON -DBUILD_LIST=dnn,ts .. && make -j4
```

* Merge a PR https://github.com/opencv/opencv/pull/11867
* Add `-DWITH_OPENCL=ON` for OpenCL support.
* Follow https://github.com/opencv/opencv/wiki/Intel's-Deep-Learning-Inference-Engine-backend to build OpenCV with Intel's Inference Engine library.

**Comment by JVision** [comment]
tensorflow reshape seems not working the same way as opencv dnn.

**Comment by dkurt** [comment]
@JVision, What do you mean by this?

We have several tests for reshape from TF:

```python
inp = tf.placeholder(tf.float32, [2, 3, 4], 'input')
shift = tf.Variable(tf.random_normal([2, 3, 4]), name='shift')
shifted = tf.add(inp, shift, name='shifted')
reshape = tf.reshape(shifted, [4, 3, 2], 'reshaped')

################################################################################
inp = tf.placeholder(tf.float32, [1, 2, 3], 'input')
reshape = tf.reshape(inp, [3, 1, 2], 'reshaped')

################################################################################
inp = tf.placeholder(tf.float32, [1, 2, 3], 'input')
bn = tf.add(inp, tf.Variable(tf.random_normal(inp.shape)))
reshape = tf.reshape(bn, [-1, 3])

################################################################################
inp = tf.placeholder(tf.float32, [None, 2, 3], 'input')
flatten = tf.reshape(inp, [-1, 2*3], 'planar')
reshaped = tf.reshape(flatten, tf.shape(inp), 'reshape')
```

Can you provide minimal working reproducer?

**Comment by shardulparab97** [comment]
@dkurt can I pick up one of the tests to get a good hold of the code?

**Comment by dkurt** [comment]
@shardulparab97, go ahead! Find any SkipTestException and try to enable it.

**Comment by i-murzov** [comment]
> Layer_Test_Halide/Convolution.Accuracy,([6, 4, 1], 5x6, 3x1, 1x1, 0x1, 1x1, false) beetle (unstable)
> Layer_Test_Halide/Convolution.Accuracy,([6, 4, 1], 5x6, 3x1, 1x1, 0x1, 2x2, true) beetle (unstable)

These two seems to be fixed by fae329a0ca6771f24a7e95c7f9a37062dbceca33

**Comment by sayoojbk** [comment]
@dkurt can I move forward working with Test_TensorFlow_layers.reshape if no one is working on this.

**Comment by dkurt** [comment]
@sayoojbk, Feel free to open a pull request for any possible test enabling. Thanks!
